partner_validate_vat/models/res_partner.py

```diff
@@ -0,0 +1,43 @@
+# coding: utf-8
+# Copyright 2017 Vauxoo (https://www.vauxoo.com) <info@vauxoo.com>
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl).
+
+from openerp import models, api, fields, _
+from openerp.exceptions import ValidationError
+
+
+class ResPartner(models.Model):
+    _inherit = 'res.partner'
+
+    vat = fields.Char(copy=False)
+
+    @api.constrains('vat', 'country_id', 'commercial_partner_id')
+    def _check_vat_uniqueness(self):
+        """Check that the partner's vat is unique among partners of the same
+        company
+        1.- Validate that the partner's vat is unique, in relation to another
+        partner, excluding those related to the commercial_parent_id field.
+        2.- Validate that the partners belong to the same company through
+        the country_id field, since two partners or more can have same vat
+        but is from different countries.
+        3.- The partners that are related have the same fields: vat,
+        commercial_parent_id, country_id
+        """
+
+        user_company = self.env.user.company_id
+
+        # Check if it belongs to the same company in relation to point 2
+        for partner in self.filtered(
+                lambda r: r.country_id == user_company.country_id and r.vat):
```

Comment:
I mean, you could use:

```python

commercial_vat = fields.Char(compute=compute_commercial_vat, store=True)

@api.depends('vat', 'commercial_partner_id')
def compute_commercial_vat(self):
   for partner in self.filtered(lambda p: not p.commercial_partner_id and p.vat):  # Detect just global partners avoiding children
      partner.commercial_vat = re.sub(r'\W+|\_', '', partner.vat.upper())  # Replace all special chars and use uppercase in order to raise the exception if a small variation is used.

_sql_constraints = [
  UNIQUE('commercial_vat', 'company_id'),
]
```

NOTE: This code is just a example and there are syntax errors, used to understand the logic.

It code is so so so faster.